File: examples/pydantic-ai/emem_mcp_geospatial_agent.py
"""Pydantic AI + emem MCP example -- structured typed answer with receipt fields.

Connects a Pydantic AI agent to the emem MCP server over Streamable HTTP
and asks for Lake Erie algal bloom hotspot data, returning a structured
typed answer.

Install:
    pip install pydantic-ai

Usage:
    export OPENAI_API_KEY="sk-..."
    python emem_mcp_geospatial_agent.py

The agent will hunt for algal bloom hotspots in Lake Erie and return a
structured answer with fields: place, event, top_cell, primary_band,
value, fact_cid, scene_url, caveats.
"""

import asyncio
import os

from pydantic_ai import Agent
from pydantic_ai.mcp import MCPServerStreamableHTTP

EMEM_MCP_URL = os.getenv("EMEM_MCP_URL", "https://emem.dev/mcp")


async def main() -> None:
    server = MCPServerStreamableHTTP(EMEM_MCP_URL)

    agent = Agent(
        "openai:gpt-4.1-mini",
        toolsets=[server],
        system_prompt=(
            "You are a geospatial evidence agent. "
            "Use emem tools to find algal bloom hotspots. "
            "Return a structured answer with these fields: "
            "place, event, top_cell, primary_band, value, fact_cid, scene_url, caveats. "
            "Use only facts that emem can support."
        ),
    )

    async with server:
        result = await agent.run(
            "Using emem, find algal bloom hotspots in Lake Erie. "
            "Return a structured answer with fields: place, event, top_cell, "
            "primary_band, value, fact_cid, scene_url, and caveats."
        )

    print(result.output)


if __name__ == "__main__":
    asyncio.run(main())

File: examples/mastra/emem_mcp_geospatial_agent.ts
// Mastra + emem MCP example — algal-bloom hotspots over Lake Erie.
//
// Wires a Mastra Agent to the emem MCP server (Streamable HTTP), lets
// the agent auto-discover all tools, and asks it to hunt for active
// algal-bloom hotspots. The agent dispatches `emem_hunt` and reports
// each hotspot with its cell64, NDVI/NDWI values, and fact_cid.
//
// Install:
//   npm install @mastra/core @mastra/mcp @ai-sdk/openai
//
// Run:
//   export OPENAI_API_KEY="sk-..."
//   npx tsx emem_mcp_geospatial_agent.ts
//
// emem itself requires no API key — reads are anonymous and
// ed25519-signed. The OpenAI key is only for the model driving the
// agent.

import { Agent } from "@mastra/core/agent";
import { MCPClient } from "@mastra/mcp";
import { openai } from "@ai-sdk/openai";

const EMEM_MCP_URL = process.env.EMEM_MCP_URL ?? "https://emem.dev/mcp";

const QUESTION =
  "Using the emem MCP tools, find the most likely active algal bloom " +
  "hotspots over Lake Erie. Call emem_hunt with event='algal_bloom' " +
  "and region='Lake Erie'. For each hotspot, report the cell64, the " +
  "primary band value, the gate band value, and the fact_cid. Do not " +
  "invent fact_cids; only quote what the tool returns. End with the " +
  "responder pubkey from the receipt.";

async function main(): Promise<void> {
  const mcp = new MCPClient({
    servers: {
      emem: {
        url: new URL(EMEM_MCP_URL),
      },
    },
  });

  const tools = await mcp.getTools();

  const agent = new Agent({
    name: "geospatial-evidence-agent",
    instructions:
      "You are an evidence-first geospatial agent. Every value you " +
      "report must come from a tool result, and you cite the " +
      "fact_cid so a reader can independently verify at /verify.",
    model: openai("gpt-4.1-mini"),
    tools,
  });

  const result = await agent.generate(QUESTION);
  console.log(result.text);

  await mcp.disconnect();
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
